language/move-prover/spec-lang/src/translate.rs

```diff
@@ -1033,16 +1058,46 @@ impl<'env, 'translator> ModuleTranslator<'env, 'translator> {
         let params = et.analyze_and_add_params(&def.signature.parameters);
         let result_type = et.translate_type(&def.signature.return_type);
         let is_public = matches!(def.visibility, PA::FunctionVisibility::Public(..));
+        let loc = et.to_loc(&def.loc);
         et.parent.parent.define_fun(
-            et.to_loc(&def.loc),
-            qsym,
+            loc.clone(),
+            qsym.clone(),
             et.parent.module_id,
             fun_id,
             is_public,
+            type_params.clone(),
+            params.clone(),
+            result_type.clone(),
+        );
+
+        // Add function as a spec fun entry as well.
+        let spec_fun_id = SpecFunId::new(self.spec_funs.len());
+        self.parent.define_spec_fun(
+            &loc,
+            qsym,
+            self.module_id,
+            spec_fun_id,
+            type_params.iter().map(|(_, ty)| ty.clone()).collect(),
+            params.iter().map(|(_, ty)| ty.clone()).collect(),
+            result_type.clone(),
+        );
+
+        // Add $ to the name so the spec version does not name clash with the move version.
```

Comment:
Why is it needed? Aren't those functions disjoint by definition? For the SpecFunEntry, we don't seem to need it.

Reply:
There will be two versions of pure functions generated, a boogie **function** for the spec version and a boogie **procedure** for the regular version. The function and the procedure need to have different names, so I add `$` to the function's name.